### ISSUE TITLE
`withZoneArena` is leaking a `Future`

### ISSUE THREAD

**Issue by blaugold** [dart-lang/ffi#107]
When the computation returns a `Future` `withZoneArena` creates a new `Future` but does not return it:
https://github.com/dart-lang/ffi/blob/41459436d576c423d3844aad7970b32d02dbd0a9/lib/src/arena.dart#L153

If the original `Future` rejects, the `Future` created in `withZoneArena` causes an unhandled exception.

**Comment by dcharkes** [comment]
Hi @blaugold,

I don't write async code every day. Could you give a minimal example?

**Comment by blaugold** [comment]
```dart
import 'dart:async';

import 'package:ffi/ffi.dart';

void main(List<String> arguments) {
  runZonedGuarded(() {
    withZoneArena(() async {
      throw 'Oops';
      return null;
    }).catchError((error) {
      print('Future error handler: $error');
    });
  }, (error, stackTrace) {
    print('Uncaught error handler: $error');
  });
}
```

This is the output:
```
➜  dart run
Uncaught error handler: Oops
Future error handler: Oops
```

**Comment by dcharkes** [comment]
I can reproduce this.

```
type 'Null' is not a subtype of type 'Future<Never>'
#0      _FutureListener.handleError (dart:async/future_impl.dart:169:20)
#1      Future._propagateToListeners.handleError (dart:async/future_impl.dart:716:47)
#2      Future._propagateToListeners (dart:async/future_impl.dart:737:24)
#3      Future._completeError (dart:async/future_impl.dart:547:5)
#4      Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart:603:7)
#5      StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:208:15)
#6      StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:116:48)
#7      _rootRun (dart:async/zone.dart:1354:13)
#8      _CustomZone.run (dart:async/zone.dart:1258:19)
#9      _CustomZone.runGuarded (dart:async/zone.dart:1162:7)
#10     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1202:23)
#11     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#12     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#13     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:120:13)
#14     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:402:11)
#15     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:426:5)
#16     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

We still release resources, but we have an uncaught error.

I'd have to think a bit more about this (what is the meaning of `Future<Never>` and why do we have it here).

**Comment by blaugold** [comment]
@dcharkes  In my example the line `return null;` is deliberate to allow the `catchError` handler to return `null`/nothing. Even though the function throws before it returns `null` it changes the signatures to `Future<Null> Function()` instead of `Future<Never> Function()`.  The `catchError` handler either has to return a value of the type of the `Future` it was added to throw or. Maybe that's what you are seeing.

**Comment by dcharkes** [comment]
We've published `1.1.2` to address this.
